mesonbuild/backend/ninjabackend.py

```diff
@@ -123,21 +123,26 @@ def get_rsp_threshold() -> int:
 NINJA_QUOTE_VAR_PAT = re.compile(r"[$ \n]")
 
 def ninja_quote(text: str, is_build_line: bool = False) -> str:
-    if is_build_line:
-        quote_re = NINJA_QUOTE_BUILD_PAT
-    else:
-        quote_re = NINJA_QUOTE_VAR_PAT
-    # Fast path for when no quoting is necessary
-    if not quote_re.search(text):
-        return text
     if '\n' in text:
         errmsg = f'''Ninja does not support newlines in rules. The content was:
 
 {text}
 
 Please report this error with a test case to the Meson bug tracker.'''
         raise MesonException(errmsg)
-    return quote_re.sub(r'$\g<0>', text)
+
+    if is_build_line:
+        if ':' in text:
+            return NINJA_QUOTE_BUILD_PAT.sub(r'$\g<0>', text)
+        quote_re = NINJA_QUOTE_BUILD_PAT
+    else:
+        quote_re = NINJA_QUOTE_VAR_PAT
+
+    if ' ' in text:
+        return quote_re.sub(r'$\g<0>', text)
+    if '$' in text:
+        return quote_re.sub(r'$\g<0>', text)
+    return text
```

Comment:
I guess the basic effect of this is to transform the "fast path" of `if not quote_re.search(text): return text` to `if ' ' in text or '$' in text or (is_build_line and ':' in text): return quote_re.sub(...)`. Do we really need to unroll it each time or can we chain some of the ifs together?

Reply:
Rewritten like you suggested.